### ISSUE TITLE
DefaultRecordMapper shouldn't apply ProxyMapper for abstract classes

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#13551]
The `ProxyMapper` currently applies early on:

```java
        // [#1470] Return a proxy if the supplied type is an interface
        // [#10071] [#11148] Primitive types are abstract! They're mapped by a ConverterProvider only later
        if (Modifier.isAbstract(type.getModifiers()) && !type.isPrimitive()) {
            delegate = new ProxyMapper();
            return;
        }
```

It automatically applies to abstract classes, not just interfaces, which cannot be proxied. Th eissue is similar to the one where we applied `ProxyMapper` to primitive types, see:

- https://github.com/jOOQ/jOOQ/issues/10071
- https://github.com/jOOQ/jOOQ/issues/11148

**Comment by lukaseder** [comment]
I don't know why we're checking for the type to be abstract, when there's a `Modifier.isInterface()` method...

**Comment by lukaseder** [comment]
The original issue report is here:
https://stackoverflow.com/q/72214794/521799

The user has applied a `Converter` to produce the type of an abstract class, and to avoid having to repeat the column expression, they just used the class literal of the abstract class (which isn't really shorter than repeating the column expression, though). This apparently used to work in the past, but now no longer does, since #11148.

I don't see a very good reason why users would use the API like this, so I'm willing to claim that the user profited from something that worked "by accident."

**Comment by mihalyr** [comment]
Hi @lukaseder I run into a breaking change when updating from `3.14.11` to the next patch version `3.14.12`

The code is basically this:

**Comment by mihalyr** [comment]
Hi @lukaseder I run into a breaking change when updating from `3.14.11` to the next patch version `3.14.12`

The code is basically this:

```java
db.select(myJSONBField).from(myTable).limit(1).fetchInto(MyResult.class);
```

The tricky part is `MyResult.class` which is defined as:

```java
@org.immutables.value.Value
@JsonDeserialize(as = ImmutableMyResult.class)
public interface MyResult {
  Person person();
  List<Task> tasks();
}
```

I've tracked down the problem to this change https://github.com/jOOQ/jOOQ/commit/08cfd28898332391a3e66cda20d3a44a9c4708d6

On `3.14.11` the `DefaultRecordMapper` will reach this point: 

```java
        // [#10071] Single-field Record1 types can be mapped if there is a ConverterProvider allowing for this mapping
        if (fields.length == 1 && Tools.converter(configuration, fields[0].getType(), type) != null) {
            delegate = new ValueTypeMapper();
            return;
        }
```

and assign the `ValueTypeMapper()` delegate correctly and everything works.

On `3.14.12` and above, the `DefaultRecordMapper` stops before that point and matches this:

```java
        // [#1470] Return a proxy if the supplied type is an interface
        // [#10071] [#11148] Primitive types are abstract! They're mapped by a ConverterProvider only later
        if (Modifier.isAbstract(type.getModifiers()) && !type.isPrimitive()) {
            delegate = new ProxyMapper();
            return;
        }
```

the `ProxyMapper()` delegate won't work here and will return `null` and my code will fail with `NullPointerExceptions`.

I could just remove the magic and map the fields by hand, but it was quite convenient to just rely on the naming convention and let JOOQ do the magic for me :) 

Basically, I wanted to raise to your attention that `ProxyMapper` will not work even for some interfaces out there as my impression from the above comments was that we want to fix this only for abstract classes and leave interfaces as-is (broken in my case).

**Comment by lukaseder** [comment]
Thanks a lot for documenting this, @mihalyr. It's kinda related, because the same change caused multiple problems, but it's not strictly about abstract classes. Yes, `ProxyMapper` doesn't work for all interfaces, although the idea here of what is supported is relatively simple.

With ad-hoc conversion having been introduced in jOOQ 3.15 for various reasons, I wonder if the reflective `DefaultRecordMapper` is still a viable approach to mapping in jOOQ. There's always some implicitness and error proneness involved, which is very foreign to the otherwise type safe jOOQ. Historically, new mapping functionality often broke something else, because of the reliance on the decision order of what mapping feature applies before what other mapping feature. It seems almost impossible to get this right.

See ad-hoc conversion: https://www.jooq.org/doc/latest/manual/sql-execution/fetching/ad-hoc-converter/

**Comment by mihalyr** [comment]
Thank you @lukaseder for your response. I am reading a complex JSON object from the DB that I want to map Jackson-like to my POJO. JOOQ ValueTypeMapper worked well for me and automatically up to 3.14.11. But I have an easy workaround, I can just use the implementation class instead of the interface and everything works as before with the latest JOOQ versions too. In, the worst case, I could supply my own converter that just uses Jackson deserialization or build my object by hand, so not a big problem at all.
